ipaserver/plugins/servicedelegation.py

```diff
@@ -212,22 +223,39 @@ def post_callback(self, ldap, completed, failed, dn, entry_attrs,
         failed[self.principal_failedattr] = {}
         failed[self.principal_failedattr][self.principal_attr] = []
         names = options.get(self.member_names[self.principal_attr], [])
-        ldap_obj = self.api.Object['service']
+        basedn = self.api.env.container_accounts + self.api.env.basedn
         if names:
             for name in names:
                 if not name:
                     continue
-                name = normalize_principal(name)
-                obj_dn = ldap_obj.get_dn(name)
+
                 try:
-                    ldap.get_entry(obj_dn, ['krbprincipalname'])
+                    n = kerberos.Principal(name, realm=self.api.env.realm)
+                except ValueError:
+                    raise errors.ValidationError(
+                        name='principal',
+                        reason=_("Malformed principal: %(error)s") % dict(
+                            error=str(e)))
+
+                if len(n.components) == 1 and (n.components[0][-1] != '$'):
```

Comment:
Please read the commit message.

Reply:
You cannot use `#is_user`, it makes no sense in the context we have here. You cannot distinguish host names and users by using that method because it only checks we have a single component.

We don't need to support users in service delegation rules, we have to add support for host principal instead. But since host principal is not a service (it is a host object) you either accept that the principal has one component (hostname itself) or do more expensive checks for existence of the host via LDAP queries. The latter is not really needed.

The only thing that prevents using this logic unconditionally are aliases for Windows-style machine accounts which consist of NetBIOS name of a machine with `$` sign appended. There are at least two users of this method: Samba and SQL Server. In case they need the delegation (S4U2Proxy is OK for them both, if needed), their service delegation rule has to contain Kerberos principal alias, not the canonical name (`cifs/...` or `MSSQLSvc/...`). In that case we must not prepend `host/..` to the single-component principal name.